drivers/ethernet/eth_smsc911x.c

```diff
@@ -0,0 +1,718 @@
+
+	if (time_out == 0) {
+		return -ETIMEDOUT;
+	}
+
+	if (smsc_mac_regread(SMSC9220_MAC_MII_DATA, data) < 0) {
+		return -1;
+	}
+
+	return 0;
+}
+
+int smsc_phy_regwrite(u8_t regoffset, u32_t data)
+{
+	u32_t val = 0;
+	u32_t phycmd = 0;
+	unsigned int time_out = REG_WRITE_TIMEOUT;
+
+	if (smsc_mac_regread(SMSC9220_MAC_MII_ACC, &val) < 0) {
+		return -1;
+	}
+
+	if (val & MAC_MII_ACC_MIIBZY) {
+		return -EBUSY;
+	}
+
+	if (smsc_mac_regwrite(SMSC9220_MAC_MII_DATA, data & 0xFFFF) < 0) {
+		return -1;
+	}
+
+	phycmd |= PHY_ADDR << 11;
+	phycmd |= (regoffset & 0x1F) << 6;
+	phycmd |= MAC_MII_ACC_WRITE;
+	phycmd |= MAC_MII_ACC_MIIBZY; /* Operation start */
+
+	if (smsc_mac_regwrite(SMSC9220_MAC_MII_ACC, phycmd) < 0) {
+		return -1;
+	}
+
+	do {
+		k_sleep(1);
+		time_out--;
+		if (smsc_mac_regread(SMSC9220_MAC_MII_ACC, &phycmd)) {
+			return -1;
+		}
+	} while (time_out != 0 && (phycmd & MAC_MII_ACC_MIIBZY));
+
+	if (time_out == 0) {
+		return -ETIMEDOUT;
+	}
+
+	return 0;
+}
+
+static int smsc_read_mac_address(u8_t *mac)
+{
+	u32_t tmp;
+	int res;
+
+	res = smsc_mac_regread(SMSC9220_MAC_ADDRL, &tmp);
+	if (res < 0) {
+		return res;
+	}
+
+	mac[0] = (u8_t)(tmp >> 0);
+	mac[1] = (u8_t)(tmp >> 8);
+	mac[2] = (u8_t)(tmp >> 16);
+	mac[3] = (u8_t)(tmp >> 24);
+
+	res = smsc_mac_regread(SMSC9220_MAC_ADDRH, &tmp);
+	if (res < 0) {
+		return res;
+	}
+
+	mac[4] = (u8_t)(tmp >> 0);
+	mac[5] = (u8_t)(tmp >> 8);
+
+	return 0;
+}
+
+static int smsc_check_id(void)
+{
+	u32_t id = SMSC9220->ID_REV;
+
+	/* If bottom and top halves of the word are the same,
+	 * the hardware is (likely) not present.
+	 */
+	if (((id >> 16) & 0xFFFF) == (id & 0xFFFF)) {
+		return -1;
+	}
+
+	switch (((id >> 16) & 0xFFFF)) {
+	case 0x9220: /* SMSC9220 on MPS2 */
+	case 0x0118: /* SMS9118 as emulated by QEMU */
+		break;
+
+	default:
+		return -1;
+	}
+
+	return 0;
+}
+
+static int smsc_soft_reset(void)
+{
+	unsigned int time_out = RESET_TIMEOUT;
+
+	SMSC9220->HW_CFG |= HW_CFG_SRST;
+
+	do {
+		k_sleep(1);
+		time_out--;
+	} while (time_out != 0 && (SMSC9220->HW_CFG & HW_CFG_SRST));
+
+	if (time_out == 0) {
+		return -1;
+	}
+
+	return 0;
+}
+
+void smsc_set_txfifo(unsigned int val)
+{
+	/* 2kb minimum, 14kb maximum */
+	if (val >= 2 && val <= 14) {
+		SMSC9220->HW_CFG = val << 16;
+	}
+}
+
+void smsc_init_irqs(void)
+{
+	SMSC9220->INT_EN = 0;
+	/* Clear all interrupts */
+	SMSC9220->INT_STS = 0xFFFFFFFF;
+	/* Polarity config which works with QEMU */
+	/* IRQ deassertion at 220 usecs and master IRQ enable */
+	SMSC9220->IRQ_CFG = 0x22000111;
+}
+
+static int smsc_check_phy(void)
+{
+	u32_t phyid1, phyid2;
+
+	if (smsc_phy_regread(SMSC9220_PHY_ID1, &phyid1)) {
+		return -1;
+	}
+	if (smsc_phy_regread(SMSC9220_PHY_ID2, &phyid2)) {
+		return -1;
+	}
+	return ((phyid1 == 0xFFFF && phyid2 == 0xFFFF) ||
+	    (phyid1 == 0x0 && phyid2 == 0x0));
+}
+
+int smsc_reset_phy(void)
+{
+	u32_t val;
+
+	if (smsc_phy_regread(SMSC9220_PHY_BCONTROL, &val)) {
+		return -1;
+	}
+
+	val |= 1 << 15;
+
```

Comment:
you can remove that line since you use val in the if ()

Reply:
> you can remove that line since you use val in the if ()

Well, I'd say it's more symmetrical (read-modify-write) and easier to read the original way.